Framework/HistogramData/inc/MantidHistogramData/Histogram.h

```diff
@@ -0,0 +1,462 @@
+  // behavior which always has Dx allocated.
+  MantidVec &dataDx() & {
+    if (!m_dx)
+      m_dx = Kernel::make_cow<HistogramDx>(m_x->size(), 0.0);
+    return m_dx.access().mutableRawData();
+  }
+  const MantidVec &dataDx() const & {
+    if (!m_dx)
+      m_dx = Kernel::make_cow<HistogramDx>(m_x->size(), 0.0);
+    return m_dx->rawData();
+  }
+  const MantidVec &readDx() const {
+    if (!m_dx)
+      m_dx = Kernel::make_cow<HistogramDx>(m_x->size(), 0.0);
+    return m_dx->rawData();
+  }
+
+private:
+  template <class TX> void initX(const TX &x);
+  template <class TY> void initY(const TY &y);
+  template <class TE> void initE(const TE &e);
+  template <class TY> void setValues(const TY &y);
+  template <class TE> void setUncertainties(const TE &e);
+  template <class T> void checkSize(const T &data) const;
+  template <class... T> bool selfAssignmentX(const T &...) { return false; }
+  template <class... T> bool selfAssignmentDx(const T &...) { return false; }
+  template <class... T> bool selfAssignmentY(const T &...) { return false; }
+  template <class... T> bool selfAssignmentE(const T &...) { return false; }
+  void switchDxToBinEdges();
+  void switchDxToPoints();
+
+  XMode m_xMode;
+};
+
+template <> MANTID_HISTOGRAMDATA_DLL void Histogram::initX(const Points &x);
+template <> MANTID_HISTOGRAMDATA_DLL void Histogram::initX(const BinEdges &x);
+template <> MANTID_HISTOGRAMDATA_DLL void Histogram::setValues(const Counts &y);
+template <>
+MANTID_HISTOGRAMDATA_DLL void Histogram::setValues(const Frequencies &y);
+template <>
+MANTID_HISTOGRAMDATA_DLL void
+Histogram::setUncertainties(const CountVariances &e);
+template <>
+MANTID_HISTOGRAMDATA_DLL void
+Histogram::setUncertainties(const CountStandardDeviations &e);
+template <>
+MANTID_HISTOGRAMDATA_DLL void
+Histogram::setUncertainties(const FrequencyVariances &e);
+template <>
+void MANTID_HISTOGRAMDATA_DLL
+Histogram::setUncertainties(const FrequencyStandardDeviations &e);
+
+template <class TX, class TY, class TE>
+Histogram::Histogram(const TX &x, const TY &y, const TE &e) {
+  initX(x);
+  initY(y);
+  initE(e);
+}
+
+template <class TY> void Histogram::initY(const TY &y) {
+  if (y)
+    setValues(y);
+}
+
+template <class TE> void Histogram::initE(const TE &e) {
+  if (e) {
+    if (!m_y)
+      throw std::logic_error("Histogram: attempted to set uncertainties for "
+                             "histogram without data");
+    setUncertainties(e);
+  } else if (m_y) {
+    setCountVariances(m_y->rawData());
+  }
+}
+
+/** Sets the Histogram's bin edges.
+
+ Any arguments that can be used for constructing a BinEdges object are allowed,
+ however, a size check ensures that the Histogram stays valid, i.e., that x and
+ y lengths are consistent. */
+template <typename... T> void Histogram::setBinEdges(T &&... data) & {
+  BinEdges edges(std::forward<T>(data)...);
+  // If there is no data changing the size is ok.
+  // if(m_y)
```

Comment:
@OwenArnold You will be wondering about this. The problem here is that `EventWorkspace` ignores the `X` length it is being initialized with, and thus requires to be changed at a later point.
I had considered allowing modification of the `X` length if there is no data, but I am not sure. I am currently working around this in the `EventWorkspace` code, but maybe it would be better to fix `EventWorkspace` instead. Let me know what you think.